src/Sarif/Core/Tool.cs

```diff
@@ -15,50 +16,46 @@ namespace Microsoft.CodeAnalysis.Sarif
         {
-            Assembly assembly = Assembly.GetEntryAssembly() ?? Assembly.GetExecutingAssembly();
+            assembly = assembly ?? Assembly.GetEntryAssembly() ?? Assembly.GetExecutingAssembly();
             string name = Path.GetFileNameWithoutExtension(assembly.Location);
-
-            Tool tool = new Tool();
-
-            // 'name'
-            tool.Name = name;
-
-            // 'version' : primary tool version.
             Version version = assembly.GetName().Version;
-            tool.Version = version.ToString();
 
-            // Synthesized semver 2.0 version required by spec
-            tool.SemanticVersion = version.Major.ToString() + "." + version.Minor.ToString() + "." + version.Build.ToString();
+            string dottedQuadFileVersion = null;
 
-            // Binary file version
             FileVersionInfo fileVersion = FileVersionInfo.GetVersionInfo(assembly.Location);
-
-            if (fileVersion.FileVersion != tool.Version)
+            if (fileVersion.FileVersion != version.ToString())
             {
-                tool.DottedQuadFileVersion = fileVersion.FileVersion;
+                dottedQuadFileVersion = fileVersion.FileVersion;
             }
 
-            tool.FullName = name + " " + tool.Version + (prereleaseInfo ?? "");
-
-            tool.Language = CultureInfo.CurrentCulture.Name;
-
-            if (!string.IsNullOrEmpty(fileVersion.Comments)) { tool.SetProperty("Comments", fileVersion.Comments); }
-            if (!string.IsNullOrEmpty(fileVersion.CompanyName)) { tool.SetProperty("CompanyName", fileVersion.CompanyName); }
-            if (!string.IsNullOrEmpty(fileVersion.ProductName)) { tool.SetProperty("ProductName", fileVersion.ProductName); }
+            Tool tool = new Tool
+            {
+                Language = CultureInfo.CurrentCulture.Name,
+                Driver = new ToolComponent
+                {
+                    Name = name,
+                    FullName = name + " " + version.ToString() + (prereleaseInfo ?? ""),
+                    Version = version.ToString(),
+                    DottedQuadFileVersion = dottedQuadFileVersion,
+                    SemanticVersion = version.Major.ToString() + "." + version.Minor.ToString() + "." + version.Build.ToString(),
+                    Properties = CreatePropertiesFromFileVersionInfo(fileVersion)
+                }
+            };
 
             return tool;
         }
 
-        public bool ShouldSerializeRulesMetadata()
-        {            
-            return this.RulesMetadata.HasAtLeastOneNonNullValue();
-        }
-
-        public bool ShouldSerializeNotificationsMetadata()
+        private static IDictionary<string, SerializedPropertyInfo> CreatePropertiesFromFileVersionInfo(FileVersionInfo fileVersion)
         {
-            return this.NotificationsMetadata.HasAtLeastOneNonNullValue();
+            var toolComponent = new ToolComponent();
```

Comment:
>ToolComponent [](start = 36, length = 13)

Can you create a property bag directly? Do you need a `toolComponent` to hold it?

Reply:
It is cumbersome to instantiate IDictionary<string, SerializedPropertyInfo> directly (I do it elsewhere in this change). This is the easier mechanism because you can use the SetProperty helper. Do you have other advice on creation? This specific changes retains the previous code pattern...

---
In reply to: [256107576](https://github.com/Microsoft/sarif-sdk/pull/1277#discussion_r256107576) [](ancestors = 256107576)